src/iceberg/arrow/arrow_error_transform_internal.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <arrow/result.h>
+#include <arrow/status.h>
+
+#include "iceberg/error.h"
+#include "iceberg/expected.h"
+
+namespace iceberg::arrow::internal {
+
+inline ErrorKind ToErrorKind(const ::arrow::Status& status) {
+  switch (status.code()) {
+    case ::arrow::StatusCode::IOError:
+      return ErrorKind::kIOError;
+    default:
+      return ErrorKind::kUnknownError;
+  }
+}
+
+#define ICEBERG_INTERNAL_ASSIGN_OR_RETURN_IMPL(result_name, lhs, rexpr, error_transform) \
```

Comment:
Perhaps `ICEBERG_ASSIGN_OR_RETURN_IMPL` or `ICEBERG_ARROW_ASSIGN_OR_RETURN_IMPL` is enough?

Reply:
ICEBERG_INTERNAL_ASSIGN_OR_RETURN_IMPL makes it easy to composite the parameter `result_name`, does it make sense to you that put ICEBERG_INTERNAL_ASSIGN_OR_RETURN_IMPL in a anonymous namespace?

Reply:
cpp-linter suggests do not use unnamed namespaces in header files, so I just leave it as is.
One thing I want to clarify is that ICEBERG_INTERNAL_ASSIGN_OR_RETURN can not use the do/while magic, because it will mess the scope of `lhs`.

Reply:
I mean we don't need to add `INTERNAL` to the macro name. We can use `ICEBERG_ARROW_` prefix.

Reply:
Renamed as suggested.